### ISSUE TITLE
GDScript: It is not possible to assign a global variable a local variable value that has the same names

### ISSUE THREAD

**Issue by viloz1** [godotengine/godot#70471]
### Godot version

4 beta 9

### System information

Windows 11

### Issue description

When having a function assign a value of a variable to a global variable of the same name it doesn't get assigned. For example, imagine we have global variable x that is assigned to "" and then have a function that takes a string x as input, and then assign the global variable x to be the value of x the value of the global variable x is not changed. This has worked before.

### Steps to reproduce

Example code:

```
extends Node

var x = ""

func _ready():
   change_x("test")
   print(x)

func change_x(x):
    x = x
```
This will print an empty string on ready.

### Minimal reproduction project

N/A

**Comment by DissonantVoid** [comment]
I think this is normal behaviour, the local x variable is shadowing the global one, to avoid this you can change the code to 
``` gdscript
    func change_x(x):
        self.x = x
``` 
by calling `self.x` you ensure that you are calling the global x variable, and it's generaly a good pratice. you can also change the name of the variable for example you can change the name of the private global variable to `_x` or the local variable to `p_x`

not sure how this worked before, it shouldn't work I think

**Comment by viloz1** [comment]
It worked in version 4 beta 3. I noticed this behavior when I came back after a break, my code broke and was fixed by changing the global variable.

Whether this is intended or not i do not know, but it does not behave in the same way.

**Comment by tinmanjuggernaut** [comment]
This should not have worked before. If it did it was probably a bug. The local variable shadows the global. 
```
var x
func change_x(x):
    x = x
```

In this function, is the left x the parameter or the global variable? How do you know it is global? It should be the local due to shadowing. That means, what you've written is "param_x = param_x". I always avoid shadowing variables. It's a very bug prone way to write code. A common convention is to prefix parameters with `p_` as mentioned.

**Comment by kleonc** [comment]
Can confirm the previous behavior (not shadowing) was a bug. Now it works as expected. For reference it was likely fixed by #69620.

Since refering to the class scope member is possible (`self.x`) I don't think there's anything more to do in here, hence closing.
